pallets/asset/src/checkpoint.rs

```diff
@@ -0,0 +1,551 @@
+//! - `advance_update_balances(ticker, updates)` advances schedules for `ticker`
+//!    and applies new balances in `updates` for the last checkpoint.
+//! - Other misc storage items as defined in `decl_storage!`.
+
+use codec::{Decode, Encode};
+use core::mem;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage,
+    dispatch::{DispatchError, DispatchResult},
+    ensure,
+    traits::{Get, UnixTime},
+};
+use frame_system::ensure_root;
+use polymesh_common_utilities::{
+    protocol_fee::{ChargeProtocolFee, ProtocolOp},
+    CommonTrait, GC_DID,
+};
+use polymesh_primitives::{
+    calendar::{CheckpointId, CheckpointSchedule},
+    IdentityId, Moment, Ticker,
+};
+use sp_runtime::traits::SaturatedConversion;
+#[cfg(feature = "std")]
+use sp_runtime::{Deserialize, Serialize};
+use sp_std::prelude::*;
+
+use crate as pallet_asset;
+use pallet_asset::Trait;
+
+type Asset<T> = pallet_asset::Module<T>;
+
+/// ID of a `StoredSchedule`.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Copy, Clone, Debug, PartialEq, Eq, Default)]
+pub struct ScheduleId(pub u64);
+
+/// One or more scheduled checkpoints in the future.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Copy, Clone, Debug, PartialEq, Eq)]
+pub struct StoredSchedule {
+    /// A series of checkpoints in the future defined by the schedule.
+    pub schedule: CheckpointSchedule,
+    /// The ID of the schedule itself.
+    /// Not to be confused for a checkpoint's ID.
+    pub id: ScheduleId,
+    /// When the checkpoint is next due to be created.
+    /// Used as a cache for more efficient sorting.
+    pub at: Moment,
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Checkpoint {
+        // --------------------- Supply / Balance storage ----------------------
+
+        /// Total supply of the token at the checkpoint.
+        ///
+        /// (ticker, checkpointId) -> total supply at given checkpoint
+        pub TotalSupply get(fn total_supply_at):
+            map hasher(blake2_128_concat) (Ticker, CheckpointId) => T::Balance;
+
+        /// Balance of a DID at a checkpoint.
+        ///
+        /// (ticker, did, checkpoint ID) -> Balance of a DID at a checkpoint
+        pub Balance get(fn balance_at_checkpoint):
+            map hasher(blake2_128_concat) (Ticker, IdentityId, CheckpointId) => T::Balance;
+
+        // ------------------------ Checkpoint storage -------------------------
+
+        /// Checkpoints created per token.
+        ///
+        /// (ticker) -> no. of checkpoints
+        pub Total get(fn total_of):
+            map hasher(blake2_128_concat) Ticker => CheckpointId;
+
+        /// Checkpoints where a DID's balance was updated.
+        /// (ticker, did) -> [checkpoint ID where user balance changed]
+        pub BalanceUpdates get(fn balance_updates):
+            map hasher(blake2_128_concat) (Ticker, IdentityId) => Vec<CheckpointId>;
```

Comment:
Could this be `double_map`?

Reply:
For the purposes of computing a DID's balance at a certain checkpoint, which is ultimately what checkpoints are for, you don't need a double map here. So I'm not sure that we need a double map here?